src/storage/query/QueryBaseProcessor.inl

```diff
@@ -524,7 +543,125 @@ kvstore::ResultCode QueryBaseProcessor<REQ, RESP>::collectEdgeProps(
             firstLoop = false;
         }
     }
-    return ret;
+}
+
+template<typename REQ, typename RESP>
+void QueryBaseProcessor<REQ, RESP>::samplingCollectEdgeProps(
+                                               kvstore::KVIterator* iter,
+                                               VertexID vId,
+                                               EdgeType edgeType,
+                                               const std::vector<PropContext>& props,
+                                               FilterContext* fcontext,
+                                               EdgeProcessor proc) {
+    auto sampler = std::make_unique<ReservoirSampling>();
+    std::vector<std::pair<std::string, std::string>> kvs;
+    kvs.resize(FLAGS_max_edge_returned_per_vertex);
+    EdgeRanking lastRank  = -1;
+    VertexID    lastDstId = 0;
+    bool        firstLoop = true;
+    int         cnt = 0;
+    bool onlyStructure = onlyStructures_[edgeType];
+    Getters getters;
+    for (; iter->valid(); iter->next()) {
+        auto key = iter->key();
+        auto val = iter->val();
+        auto rank = NebulaKeyUtils::getRank(key);
+        auto dstId = NebulaKeyUtils::getDstId(key);
+        if (!firstLoop && rank == lastRank && lastDstId == dstId) {
+            VLOG(3) << "Only get the latest version for each edge.";
+            continue;
+        }
+        lastRank = rank;
+        lastDstId = dstId;
+
+        auto index = sampler->sampling();
+        if (index >= 0 && index < FLAGS_max_edge_returned_per_vertex) {
+            kvs[index] = std::make_pair(key.str(), val.str());
+            ++cnt;
+        }
+    }
+
+    if (cnt > FLAGS_max_edge_returned_per_vertex) {
+        cnt = FLAGS_max_edge_returned_per_vertex;
+    }
+
+    for (auto i = 0; i < cnt; ++i) {
```

Comment:
Most of the code in this loop is the same as in collectEdgeProps, so it is best to abstract a common method.

Reply:
This is a big loop, is it a good idea to make it a function call?